wrouter/route_params.go

```diff
@@ -18,120 +18,98 @@ import (
-		return &requestParamPermsImpl{
-			pathParamPerms: newForbiddenParamPerms(forbiddenParams...),
-		}
+	return RouteParamPermsParam(&requestParamPermsImpl{
+		pathParamPerms: newForbiddenParamPerms(forbiddenParams...),
 	})
 }
 
 func SafeQueryParams(safeParams ...string) RouteParam {
-	return requestParamPermsFunc(func() RouteParamPerms {
-		return &requestParamPermsImpl{
-			queryParamPerms: newSafeParamPerms(safeParams...),
-		}
+	return RouteParamPermsParam(&requestParamPermsImpl{
+		queryParamPerms: newSafeParamPerms(safeParams...),
 	})
 }
 
 func ForbiddenQueryParams(forbiddenParams ...string) RouteParam {
-	return requestParamPermsFunc(func() RouteParamPerms {
-		return &requestParamPermsImpl{
-			queryParamPerms: newForbiddenParamPerms(forbiddenParams...),
-		}
+	return RouteParamPermsParam(&requestParamPermsImpl{
+		queryParamPerms: newForbiddenParamPerms(forbiddenParams...),
 	})
 }
 
 func SafeHeaderParams(safeParams ...string) RouteParam {
-	return requestParamPermsFunc(func() RouteParamPerms {
-		return &requestParamPermsImpl{
-			headerParamPerms: newSafeParamPerms(safeParams...),
-		}
+	return RouteParamPermsParam(&requestParamPermsImpl{
+		headerParamPerms: newSafeParamPerms(safeParams...),
 	})
 }
 
 func ForbiddenHeaderParams(forbiddenParams ...string) RouteParam {
-	return requestParamPermsFunc(func() RouteParamPerms {
-		return &requestParamPermsImpl{
-			headerParamPerms: newForbiddenParamPerms(forbiddenParams...),
-		}
+	return RouteParamPermsParam(&requestParamPermsImpl{
+		headerParamPerms: newForbiddenParamPerms(forbiddenParams...),
 	})
 }
 
-func toRequestParamPerms(params []RouteParam) RouteParamPerms {
-	var pathParamPerms []ParamPerms
-	var queryParamPerms []ParamPerms
-	var headerParamPerms []ParamPerms
-
-	for _, p := range params {
-		if p == nil {
-			continue
-		}
-
-		perms := p.perms()
-		if perms != nil {
-			pathParamPerms = append(pathParamPerms, perms.PathParamPerms())
-			queryParamPerms = append(queryParamPerms, perms.QueryParamPerms())
-			headerParamPerms = append(headerParamPerms, perms.HeaderParamPerms())
-		}
-	}
-
-	return &requestParamPermsImpl{
-		pathParamPerms:   NewCombinedParamPerms(pathParamPerms...),
-		queryParamPerms:  NewCombinedParamPerms(queryParamPerms...),
-		headerParamPerms: NewCombinedParamPerms(headerParamPerms...),
-	}
-}
-
-type requestMetricTagsFunc func() metrics.Tags
-
-func (f requestMetricTagsFunc) perms() RouteParamPerms {
-	return nil
-}
-
-func (f requestMetricTagsFunc) metricTags() metrics.Tags {
-	return f()
-}
-
 func MetricTags(tags metrics.Tags) RouteParam {
-	return requestMetricTagsFunc(func() metrics.Tags {
-		return tags
+	return routeParamFunc(func(b *routeParamBuilder) error {
+		b.metricTags = tags
+		return nil
 	})
 }
 
-func toMetricTags(params []RouteParam) metrics.Tags {
-	var tags metrics.Tags
-	for _, param := range params {
-		tags = append(tags, param.metricTags()...)
+func toRequestParamPerms(b *routeParamBuilder) RouteParamPerms {
```

Comment:
Since `toRequestParamPerms` and `toMetricTags` now both operate on `*routeParamBuilder`, let's have these functions be functions on the `*routeParamBuilder` type and move their definition next to the definition for the builder type.

Reply:
done